File: rust/bridge/shared/src/net/cdsi.rs
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

use std::convert::TryInto as _;

use libsignal_bridge_macros::{bridge_fn, bridge_io};
use libsignal_bridge_types::net::cdsi::{CdsiLookup, LookupRequest};
use libsignal_bridge_types::net::{ConnectionManager, TokioAsyncContext};
use libsignal_core::E164;
use libsignal_net::auth::Auth;
use libsignal_net::cdsi::{self, AciAndAccessKey, LookupResponse};
use libsignal_protocol::{Aci, SignalProtocolError};

use crate::support::*;
use crate::*;

bridge_handle_fns!(LookupRequest, clone = false);

#[bridge_fn]
fn LookupRequest_new() -> LookupRequest {
    LookupRequest::default()
}

#[bridge_fn]
fn LookupRequest_addE164(request: &LookupRequest, e164: E164) {
    request.lock().new_e164s.push(e164)
}

#[bridge_fn]
fn LookupRequest_addPreviousE164(request: &LookupRequest, e164: E164) {
    request.lock().prev_e164s.push(e164)
}

#[bridge_fn]
fn LookupRequest_setToken(request: &LookupRequest, token: &[u8]) {
    request.lock().token = token.into();
}

#[bridge_fn]
fn LookupRequest_addAciAndAccessKey(
    request: &LookupRequest,
    aci: Aci,
    access_key: &[u8],
) -> Result<(), SignalProtocolError> {
    let access_key = access_key
        .try_into()
        .map_err(|_: std::array::TryFromSliceError| {
            SignalProtocolError::InvalidArgument("access_key has wrong number of bytes".to_string())
        })?;
    request
        .lock()
        .acis_and_access_keys
        .push(AciAndAccessKey { aci, access_key });
    Ok(())
}

bridge_handle_fns!(CdsiLookup, clone = false);

#[bridge_io(TokioAsyncContext)]
async fn CdsiLookup_new(
    connection_manager: &ConnectionManager,
    username: String,
    password: String,
    request: &LookupRequest,
) -> Result<CdsiLookup, cdsi::LookupError> {
    let request = std::mem::take(&mut *request.lock());
    let auth = Auth { username, password };

    CdsiLookup::new(connection_manager, auth, request).await
}

#[bridge_io(TokioAsyncContext)]
async fn CdsiLookup_new_routes(
    connection_manager: &ConnectionManager,
    username: String,
    password: String,
    request: &LookupRequest,
) -> Result<CdsiLookup, cdsi::LookupError> {
    let request = std::mem::take(&mut *request.lock());
    let auth = Auth { username, password };

    CdsiLookup::new_routes(connection_manager, auth, request).await
}

#[bridge_fn]
fn CdsiLookup_token(lookup: &CdsiLookup) -> &[u8] {
    &lookup.token.0
}

#[bridge_io(TokioAsyncContext)]
async fn CdsiLookup_complete(lookup: &CdsiLookup) -> Result<LookupResponse, cdsi::LookupError> {
    lookup
        .take_remaining()
        .expect("not completed yet")
        .collect()
        .await
}

File: node/ts/net/CDSI.ts
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import type { ReadonlyDeep } from 'type-fest';
import * as Native from '../Native.js';
import { Aci } from '../Address.js';
import { Buffer } from 'node:buffer';
import { TokioAsyncContext, ServiceAuth } from '../net.js';
import { newNativeHandle } from '../internal.js';

export type CDSRequestOptionsType = {
  e164s: Array<string>;
  acisAndAccessKeys: Array<{ aci: string; accessKey: string }>;
  abortSignal?: AbortSignal;
};

export type CDSResponseEntryType<Aci, Pni> = {
  aci: Aci | undefined;
  pni: Pni | undefined;
};

export type CDSResponseEntries<Aci, Pni> = Map<
  string,
  CDSResponseEntryType<Aci, Pni>
>;

export interface CDSResponseType<Aci, Pni> {
  entries: CDSResponseEntries<Aci, Pni>;
  debugPermitsUsed: number;
}

export async function cdsiLookup(
  {
    asyncContext,
    connectionManager,
  }: Readonly<{
    asyncContext: TokioAsyncContext;
    connectionManager: Native.Wrapper<Native.ConnectionManager>;
  }>,
  { username, password }: Readonly<ServiceAuth>,
  { e164s, acisAndAccessKeys, abortSignal }: ReadonlyDeep<CDSRequestOptionsType>
): Promise<CDSResponseType<string, string>> {
  const request = newNativeHandle(Native.LookupRequest_new());
  e164s.forEach((e164) => {
    Native.LookupRequest_addE164(request, e164);
  });

  acisAndAccessKeys.forEach(({ aci: aciStr, accessKey: accessKeyStr }) => {
    Native.LookupRequest_addAciAndAccessKey(
      request,
      Aci.parseFromServiceIdString(aciStr).getServiceIdFixedWidthBinary(),
      Buffer.from(accessKeyStr, 'base64')
    );
  });

  const lookup = await asyncContext.makeCancellable(
    abortSignal,
    Native.CdsiLookup_new(
      asyncContext,
      connectionManager,
      username,
      password,
      request
    )
  );
  return await asyncContext.makeCancellable(
    abortSignal,
    Native.CdsiLookup_complete(asyncContext, newNativeHandle(lookup))
  );
}
